### PR TITLE
Lite: Support binding non-bphwc4 metal buffer to tensor

### PR DESCRIPTION
**Pull request by jongkweh**
## Purpose
This is to support binding non-bphwc4 metal buffer to tensor. 

## Background
When binding metal buffer to tensor, the current implementation assumes the metal buffer to be in bphwc4 format.  
In order to bind a metal buffer to tensor, the user needs to convert the metal buffer to bphwc4 format. (for CPU inputs, this is taken care inside the metal delegate). PHWC4 is a unique format used inside TfLite so i think this should be taken care inside metal delegate just like the CPU input buffers.  

## Considerations
If we forcefully convert every binded metal buffer to phwc4 format, users that already are already using `TFLGpuDelegateBindMetalBufferToTensor` function with custom phwc4 might be affected. 
So there needs to be a way to check whether the input metal buffer is in phwc4 format. 
I guess you could add a parameter for `TFLGpuDelegateBindMetalBufferToTensor` the function but without adding extra parameter, one way we can estimate is by the metal buffer length (but  it might not be enough).

### CODE REVIEW

**Comment by gbaned** [comment]
Hi @terryheo Can you please review this PR ? Thank you!

File: tensorflow/lite/delegates/gpu/metal_delegate.mm [hunk 1]
```diff
@@ -211,7 +211,12 @@ kernel void ComputeFunction(device int* output_buffer [[buffer(0)]],
     for (auto& input : graph_inputs_) {
       if (input.tensor_id == tensor_index) {
         if (in_out_tensors_[input.id]->GetBufferHandle() != buffer) {
-          RETURN_IF_ERROR(in_out_tensors_[input.id]->SetBufferHandle(buffer));
+          if ([in_out_tensors_[input.id]->GetBufferHandle() length] == [buffer length]) {
+            RETURN_IF_ERROR(in_out_tensors_[input.id]->SetBufferHandle(buffer));
+            input.is_bphwc4 = true;
+          } else {
+            in_out_bhwc_f32_buffers_[input.id] = buffer;
```
Comment (terryheo) [inline tensorflow/lite/delegates/gpu/metal_delegate.mm:218]: What's the use case of non phwc4 cases? How can we test it?

File: tensorflow/lite/delegates/gpu/metal_delegate.mm [hunk 2]
```diff
@@ -221,7 +226,12 @@ kernel void ComputeFunction(device int* output_buffer [[buffer(0)]],
     for (auto& output : graph_outputs_) {
       if (output.tensor_id == tensor_index) {
         if (in_out_tensors_[output.id]->GetBufferHandle() != buffer) {
-          RETURN_IF_ERROR(in_out_tensors_[output.id]->SetBufferHandle(buffer));
+          if ([in_out_tensors_[output.id]->GetBufferHandle() length] == [buffer length]) {
+            RETURN_IF_ERROR(in_out_tensors_[output.id]->SetBufferHandle(buffer));
+            output.is_bphwc4 = true;
+          } else {
+            in_out_bhwc_f32_buffers_[output.id] = buffer;
```
Comment (terryheo) [inline tensorflow/lite/delegates/gpu/metal_delegate.mm:233]: Does this work? I think output buffer is usually created by delegate. I'm not sure how non phwc4 output will work.

File: tensorflow/lite/delegates/gpu/metal_delegate.mm (hunk 1 above)
Comment (jongkweh) [reply tensorflow/lite/delegates/gpu/metal_delegate.mm:218]: This is the case when we use  `TFLGpuDelegateBindMetalBufferToTensor`. Like any other cpu inputs, user might bind a non-phwc4 buffer. If the user binds non-phwc4 buffer (just like any other cpu inputs), the original code won't work. The user has to always add a conversion code just like `TFLBufferConvert` . Note that in gl_delegate, they actually have a code that checks where it is pbwc4 or not for external buffer (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/delegates/gpu/gl_delegate.cc#L195)

File: tensorflow/lite/delegates/gpu/metal_delegate.mm (hunk 2 above)
Comment (jongkweh) [reply tensorflow/lite/delegates/gpu/metal_delegate.mm:233]: Again, this the case when you use `TFLGpuDelegateBindMetalBufferToTensor`. Using this method, you can bind external output buffer.

Comment (jongkweh) [reply tensorflow/lite/delegates/gpu/metal_delegate.mm:233]: If assigning to `in_out_bhwc_f32_buffers_` non_pbwc4 will be outputed using this code here. https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/delegates/gpu/metal_delegate.mm#L522

Comment (gbaned) [reply tensorflow/lite/delegates/gpu/metal_delegate.mm:233]: Hi @terryheo Can you please assist on above comments from @jongkweh. Thank you!

File: tensorflow/lite/delegates/gpu/metal_delegate.mm [hunk 3]
```diff
@@ -631,6 +644,7 @@ void SetCommandBuffer(id<MTLCommandBuffer> command_buffer) {
     int64_t tensor_id;
     BHWC shape;
     bool set_externally;  // a user fills/retrieves data on this MTLBuffer buffer
+    bool is_bphwc4;
```
Comment (terryheo) [inline tensorflow/lite/delegates/gpu/metal_delegate.mm:647]: plz add some comment

File: tensorflow/lite/delegates/gpu/metal_delegate.mm (hunk 1 above)
Comment (terryheo) [reply tensorflow/lite/delegates/gpu/metal_delegate.mm:218]: I see. Could you add some comment for "in_out_bhwc_f32_buffers_[input.id] = buffer;" case ?

**Comment by gbaned** [comment]
Hi @jongkweh Can you please check @terryheo's comments and keep us posted ? Thank you!

**Pull request closed without merging**